### ISSUE TITLE
Multiword arguments

### ISSUE THREAD

**Issue by ogallagher** [ogallagher/terry#17]
Terry currently does not currently support multiword arguments (“two hundred three”, type “hello world”), but it will have to soon.

**Comment by ogallagher** [comment]
A few decisions and considerations by example:

Given the raw instruction: 

`type in the search bar in the search bar in popcorn`

and the pattern to match:

`type @$string) ?in) @wwidget)`

... currently the instruction parser removes all trivial words before continuing, so the instruction would become: 

`type in search bar in search bar in popcorn`

This will need to change, because if we assume for example that “in the search bar” is a string argument, “the” should not be removed. The trivial check should therefore not happen as early in the language parsing pipeline as the instruction parser, but rather when instruction possibilities are attempting resolution.

Then, the following possibilities arise after resolving some tokens:

```
A. type “in search bar in search bar in popcorn” X (ran out of tokens, waiting for “in” or @wwidget)
B. type “in search bar in search bar” in X (expecting @wwidget, got “popcorn”)
C. type “in search bar in” search bar (“in popcorn” ignored)
D. type “in search bar” in search bar (“in popcorn” ignored)
E. type “in” search bar (“in search bar in popcorn” ignored)
```

In the cases of C, D, and E, when a possibility could be complete but there are more tokens in the transcription, that possibility should be conserved. I’ll need to double-check if this behavior is currently supported, but I think Terry currently would throw away C-E, try to resolve A and B, and end up with no possible language mappings. If this is the case, I need to change it so all possible candidates (including those ignoring extra tokens) are then compared finally.

So, given that C, D, and E are all possible matches for the mapping with the pattern `type @$string) ?in) @wwidget)` specified above, __D__ would be chosen as the best candidate because it has the most tokens (C=4, D=5, E=4). This final choice by token count also needs to be implemented.

In conclusion Terry needs the following changes to support multiword/multitoken arguments:

- [x] move trivial check further in the pipeline
- [x] if leaf PatternNode is an argument, always include itself as a possible follower and use Arg.getArgValue() to check if the next token addition to the argument is valid
- [x] conserve complete possibilities that ignore extra tokens
- [x] use token count as the final deciding factor between candidates

**Comment by ogallagher** [comment]
I took a first pass at the second task, and ended up modifying `InstructionPossibility.resolve(token)` to, in the case that the leaf is an argument, attempt appending the new token to it as well as checking against follower nodes. If the token append is successful, the leaf node is added to the back of the `leaves` list as a remaining possibility.

**Comment by ogallagher** [comment]
Though I’ve not tested the changes yet, trivial tokens should now be ignored for resolution against keywords, but kept for resolution against args.

**Comment by ogallagher** [comment]
Possibilities that could be complete while more tokens are available are now kept, but the instruction parser does not yet support the final decision by token count between multiple candidates.

EDIT: Now the decision should be made. Time to test and fix!

**Comment by ogallagher** [comment]
I am modifying the candidate decision; instead of picking the instruction candidate with the highest _token_ count, pick the one with the highest _argument_ count. If there’s a tie for number of resolved arguments, pick the one with the highest count of tokens within arguments.

For example:
`str=“the”, num=“ten”`
__`str=“the end”, num=“ten”`__ this is the best candidate, with 3 args using 3 tokens

Also, followers cloned from leaf followers as a result of multiword arguments need to also be leaves.

**Comment by ogallagher** [comment]
Both of the above modifications seem to work correctly now. The last errors had to do with leaf followers being removed from the instruction possibility’s leaf list before the parent node was cloned, thus losing the follower leaves to be cloned.
